src/transformers/modeling_utils.py

```diff
@@ -737,20 +733,17 @@ def _load_state_dict_into_meta_model(
             if not is_safetensors:
                 offload_index = offload_weight(param, param_name, offload_folder, offload_index)
         elif param_device == "cpu" and state_dict_index is not None:
-            state_dict_index = offload_weight(param, param_name, state_dict_folder, state_dict_index)
-        elif not is_quantized:
-            # For backward compatibility with older versions of `accelerate`
+            state_dict_index = offload_weight(param, param_name, model, state_dict_folder, state_dict_index)
+        elif (
+            quantizer is None
+            or (not quantizer.requires_parameters_quantization)
+            or (not quantizer.check_quantized_param(model, param, param_name, state_dict))
+        ):
+            # For backward compatibility with older versions of `accelerate` and for non-quantized params
             set_module_tensor_to_device(model, param_name, param_device, **set_module_kwargs)
         else:
-            if param.dtype == torch.int8 and param_name.replace("weight", "SCB") in state_dict.keys():
-                fp16_statistics = state_dict[param_name.replace("weight", "SCB")]
-            else:
-                fp16_statistics = None
-
-            if "SCB" not in param_name:
-                set_module_quantized_tensor_to_device(
-                    model, param_name, param_device, value=param, fp16_statistics=fp16_statistics
-                )
+            quantizer.create_quantized_param(model, param, param_name, param_device, state_dict)
+            # TODO: consider removing used param_parts from state_dict before return
```

Comment:
I did that by passing `unexpected keys` to `quantizer.create_quantized_param()`.  not messing ith state dict.
this allows to delete method `HFQuantizer.update_mismatched_keys()`. One fewer call to quantizer.

Reply:
This would be even prettier if bnb auxiliary elements were buffers. But it was not the case for 8bit and I also copied that approach to 4bit. Perhaps this will get refactored some day on bnb side.

Reply:
I wonder, why `unexpected_keys` becomes a list? it would be more efficient to keep it as a set. @SunMarc